File: android/app/src/main/java/io/flutter/plugins/com/safe_pulse/SmsReceiver.java
package io.flutter.plugins.com.example.safe_pulse;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.os.IBinder;
import android.telephony.SmsMessage;
import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;
import io.flutter.embedding.android.FlutterActivity;

public class SmsReceiver extends BroadcastReceiver {
    private static final String SMS_RECEIVED_ACTION = "sms-received";

    @Override
    public void onReceive(Context context, Intent intent) {
        if (intent.getAction() != null && intent.getAction().equals("android.provider.Telephony.SMS_RECEIVED")) {
            Bundle bundle = intent.getExtras();
            if (bundle != null) {
                try {
                    Object[] pdus = (Object[]) bundle.get("pdus");
                    if (pdus != null) {
                        for (Object pdu : pdus) {
                            SmsMessage message;
                            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                                String format = bundle.getString("format");
                                message = SmsMessage.createFromPdu((byte[]) pdu, format);
                            } else {
                                message = SmsMessage.createFromPdu((byte[]) pdu);
                            }

                            String sender = message.getOriginatingAddress();
                            String body = message.getMessageBody();

                            // Broadcast to Flutter
                            Intent localIntent = new Intent(SMS_RECEIVED_ACTION);
                            localIntent.putExtra("sender", sender);
                            localIntent.putExtra("message", body);
                            context.sendBroadcast(localIntent);

                            // Start the SMS service
                            Intent serviceIntent = new Intent(context, SmsService.class);
                            serviceIntent.putExtra("sender", sender);
                            serviceIntent.putExtra("message", body);
                            
                            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                                context.startForegroundService(serviceIntent);
                            } else {
                                context.startService(serviceIntent);
                            }
                        }
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }
    }

    public static class SmsService extends Service {
        private static final String CHANNEL_ID = "SmsServiceChannel";
        private static final int NOTIFICATION_ID = 123;

        @Override
        public void onCreate() {
            super.onCreate();
            createNotificationChannel();
        }

        @Override
        public int onStartCommand(Intent intent, int flags, int startId) {
            String sender = intent.getStringExtra("sender");
            String message = intent.getStringExtra("message");

            // Create notification
            Intent notificationIntent = new Intent(this, FlutterActivity.class);
            notificationIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
            PendingIntent pendingIntent = PendingIntent.getActivity(
                this,
                0,
                notificationIntent,
                PendingIntent.FLAG_IMMUTABLE
            );

            Notification notification = new NotificationCompat.Builder(this, CHANNEL_ID)
                .setContentTitle("New SMS from " + sender)
                .setContentText(message)
                .setSmallIcon(android.R.drawable.ic_dialog_email) // Replace with your own icon
                .setContentIntent(pendingIntent)
                .setAutoCancel(true)
                .build();

            startForeground(NOTIFICATION_ID, notification);

            return START_NOT_STICKY;
        }

        private void createNotificationChannel() {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                NotificationChannel serviceChannel = new NotificationChannel(
                    CHANNEL_ID,
                    "SMS Service Channel",
                    NotificationManager.IMPORTANCE_DEFAULT
                );
                NotificationManager manager = getSystemService(NotificationManager.class);
                manager.createNotificationChannel(serviceChannel);
            }
        }

        @Nullable
        @Override
        public IBinder onBind(Intent intent) {
            return null;
        }
    }
}

File: android/app/src/main/java/io/flutter/plugins/com/safe_pulse/SmsReceiver.kt
package io.flutter.plugins.com.safe_pulse

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.telephony.SmsMessage
import android.util.Log

class SmsReceiver : BroadcastReceiver() {
    companion object {
        private const val TAG = "SMS_RECEIVER"
        private const val SMS_RECEIVED_ACTION = "android.provider.Telephony.SMS_RECEIVED"
        private const val FLUTTER_ACTION = "sms-received"
        private const val PDU_TYPE = "pdus"
    }

    override fun onReceive(context: Context, intent: Intent) {
        Log.d(TAG, "SMS intent received with action: ${intent.action}")
        
        if (intent.action == SMS_RECEIVED_ACTION) {
            val bundle = intent.extras ?: run {
                Log.w(TAG, "No extras found in SMS intent")
                return
            }
            processSmsIntent(context, bundle)
        }
    }

    private fun processSmsIntent(context: Context, bundle: Bundle) {
        try {
            val pdus = getPdusFromBundle(bundle)
            if (pdus.isEmpty()) {
                Log.w(TAG, "No PDUs found in bundle")
                return
            }

            val messages = parseMessagesFromPdus(bundle, pdus)
            messages.forEach { message ->
                processSingleMessage(context, message)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error processing SMS", e)
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun getPdusFromBundle(bundle: Bundle): Array<ByteArray> {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            bundle.getParcelableArray(PDU_TYPE, ByteArray::class.java)?.let {
                Array(it.size) { i -> it[i] as ByteArray }
            } ?: emptyArray()
        } else {
            @Suppress("DEPRECATION")
            (bundle.get(PDU_TYPE) as? Array<*>)?.filterIsInstance<ByteArray>()?.toTypedArray()
                ?: emptyArray()
        }
    }

    private fun parseMessagesFromPdus(bundle: Bundle, pdus: Array<ByteArray>): List<SmsMessage> {
        return pdus.mapNotNull { pdu ->
            try {
                val format = bundle.getString("format")
                SmsMessage.createFromPdu(pdu, format)
            } catch (e: Exception) {
                Log.e(TAG, "Error creating SmsMessage from PDU", e)
                null
            }
        }
    }

    private fun processSingleMessage(context: Context, message: SmsMessage) {
        val sender = message.originatingAddress ?: "unknown"
        val body = message.messageBody ?: "no content"
        
        Log.d(TAG, "Processing SMS from $sender: ${body.take(30)}...")

        forwardToMainActivity(context, sender, body)
    }

    private fun forwardToMainActivity(context: Context, sender: String, body: String) {
        try {
            val intent = Intent().apply {
                setClassName(context, "io.flutter.plugins.com.safe_pulse.MainActivity")
                action = FLUTTER_ACTION
                putExtra("sender", sender)
                putExtra("message", body)
                flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_SINGLE_TOP
            }
            context.startActivity(intent)
            Log.d(TAG, "SMS forwarded to MainActivity")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to forward SMS to MainActivity", e)
        }
    }
}